### PR TITLE
Add support for restoring a DECCTR color table report

### LINKED ISSUE #13132: Add support for the VT525 restore color table sequence
# Description of the new feature/enhancement

The VT525 terminal had a pair of sequences for saving and restoring terminal state: the `DECRQTSR` sequence could be used to request a state report from the terminal which an application could save; and then the `DECRSTS` sequences would be used to restore that saved data to the terminal, returning it to the state it was originally.

There are two kinds of reports supported by these sequences:
1. the "terminal state report", which contains most of the settable attributes, excluding large items like the DRCS fonts, and macro definitions.
2. the "color table report", which contains the color definitions for all of the palette entries in the color table.

What I'm proposing is that we add support for the color table report, but only the restore operation (for now).

The one reason I'd like to add this is so we have standard-compliant way to alter the palette. The second reason is that it will give me a chance to introduce the color conversion routines that will eventually be required for sixel.

# Proposed technical implementation details (optional)

`DECRSTS` is a `DCS` sequence, which can't pass through conpty, so for now this would only work on conhost, or using the experimental passthrough mode in Windows Terminal. That said, I am planning to try and add support for `DCS` over a regular conpty connection at some point (at least for simple cases like this). But I'm not proposing to do that now.

### PR DESCRIPTION
**Pull request by j4james**
This PR introduces the framework for the `DECRSTS` sequence which is
used to restore terminal state reports. But to start with, I've just
implemented the `DECCTR` color table report, which provides a way for
applications to alter the terminal's color scheme.

## PR Checklist
* [x] Closes #13132
* [x] CLA signed.
* [x] Tests added/passed
* [ ] Documentation updated.
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

I've added the functions for parsing DEC RGB and HLS color formats into
the `Utils` class, where we've got all our other color parsing routines,
since this functionality will eventually be needed in other VT protocols
like Sixel and ReGIS.

Since `DECRSTS` is a `DCS` sequence, this only works in conhost for now,
or when using the experimental passthrough mode in Windows Terminal.

## Validation Steps Performed

I've added a number of unit tests to check that the `DECCTR` report is
being interpreted as expected. This includes various edge cases (e.g.
omitted and out-of-range parameters), which I have confirmed to match
the color parsing on a real VT240 terminal.

### CODE REVIEW

**Review by lhecker — APPROVED** [review]
I think I realize now, that making `VTInt` a signed integer might've been a mistake... Thankfully it should be easier to change that since it's just a alias now.

If we need these two new color functions to be performant, we could use a LUT for mapping 0-100 to 0-255. That wouldn't just work for `ColorFromRGB100` but also for `ColorFromHLS` which can be written with integer arithmetic only (without integer division).

**Comment by j4james** [comment]
> I think I realize now, that making `VTInt` a signed integer might've been a mistake...

Why's that? I actually thought a signed int was the right choice. As an example, if you're working with VT coordinates, a typical calculation would be to add or subtract an offset, and then clamp the result within the active margins or buffer range. If that intermediate offset goes negative on a signed int that's not a problem - it's still going to clamp correctly. But if it's unsigned, it's going to wrap, and you're going to end up with the wrong result.

> If we need these two new color functions to be performant, we could use a LUT for mapping 0-100 to 0-255.

I like that idea. I am keen to get these functions as fast as possible, mostly for when they're used in sixel. Although HLS colors aren't likely to be used much, so the HLS one is less critical.

> That wouldn't just work for `ColorFromRGB100` but also for `ColorFromHLS` which can be written with integer arithmetic only (without integer division).

I've done some initial experiments with a LUT, and it's a big improvement for `ColorFromRGB100`, but in `ColorFromHLS` it actually seems somewhat slower. It could just my badly written code, but I went with the floating point implementation in the first place because I found it faster than my integer version, and I suspect the LUT doesn't make up for that loss.

I'll play around some more and see if I can improve things, otherwise I'll just do the LUT for `ColorFromRGB100`.

**Comment by lhecker** [comment]
> Why's that? I actually thought a signed int was the right choice.

It's just... It kind of feels wrong to me at times. For instance your code does `std::min(r, 100)`, which 100% exactly what I'd do myself. But shouldn't it be `std::clamp(r, 0, 100)`? Because if `r` was negative, the computation would result in a negative number and thus in a red value of `255` instead of `0`. We know it works, because `VTInt`s are never negative.
So I've been thinking whether it would make sense to make `VTInt` a `uint32_t` or maybe even a `uint16_t`. The former would contractually guarantee that negative numbers can't exist and the latter would even make it possible to freely convert it to both `uint32_t` or `int32_t` without any AuditCheck naggings. Hmm... 🤔
But it's not that important really...

> I've done some initial experiments with a LUT, and it's a big improvement for `ColorFromRGB100`, but in `ColorFromHLS` it actually seems somewhat slower.

Oh nice! I'm glad it works in practice! BTW here's a SO answer with a low-level HLS-RGB routine: https://stackoverflow.com/a/45450486
(That post is [CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/) so we can easily use it here if we choose to do so.)

**Comment by j4james** [comment]
> So I've been thinking whether it would make sense to make `VTInt` a `uint32_t` or maybe even a `uint16_t`. The former would contractually guarantee that negative numbers can't exist and the latter would even make it possible to freely convert it to both `uint32_t` or `int32_t` without any AuditCheck naggings.

I see your point. My main concern was that we'd find ourselves having to cast to `int` in a bunch of places, but maybe using `uint16_t` would make that less of a problem.

> BTW here's a SO answer with a low-level HLS-RGB routine: https://stackoverflow.com/a/45450486

They're relying on all the inputs being in the range 0-255, which allows them to do a bunch of bit shifting and masking that isn't applicable in our case. And then there's stuff like using a branch in place of `abs` which seems slower to me, and the `switch` in places of the final conditions didn't show any obvious benefit in our code as far as I could tell (it may be better for them because they could use a shift for the division).

So for now I think I'm happy to stick with the current HLS implementation, and just use your LUT idea in the RGB conversion. You're welcome to take a stab at optimising this further after it's merged, although as I said, the HLS routine isn't that critical.

**Comment by lhecker** [comment]
I agree. There's no need to overcomplicate it if it isn't used much. 🙂

BTW, I'm a bit surprised the `gsl::at` calls inside a `noexcept` function don't trigger the AuditMode checks. 🤔

**Comment by j4james** [comment]
> BTW, I'm a bit surprised the `gsl::at` calls inside a `noexcept` function don't trigger the AuditMode checks. 🤔

Oh wow. I've always assumed `gsl::at` was an unchecked `operator[]` with the warnings suppressed, in the same vein as `gsl::narrow_cast`. Otherwise why wouldn't we just use the checked `std::array::at` method? That at least throws an exception, whereas `gsl::at` appears to be an app killer. And I guess that's why the audit doesn't complain.

Could I just use an `operator[]` with a `pragma` to suppress the warning? I really don't want an unnecessary boundary check there and the optimizer doesn't seem able to avoid it.

**Comment by DHowett** [comment]
> Could I just use an `operator[]` with a `pragma` to suppress the warning? I really don't want an unnecessary boundary check there and the optimizer doesn't seem able to avoid it.

Eh, honestly that's what `til::at` is for. It's the "Audit-safe" "I know what I am doing" indexer.

Yes, creating new tools to work around our old tools being poor tools is the name of the game here. 😁 At least `til::at`'s call sites are hand-auditable!

**Comment by j4james** [comment]
OK, I think this is it. I've switched to `til::at`, and I'm also using `std::min<unsigned>` to make absolutely sure the parameters will be in range in the unlikely even that someone passes in a negative number.

**Review by carlos-zamora — APPROVED** [review]
LOVE IT! Thanks!

**Comment by carlos-zamora** [comment]
@msftbot make sure @DHowett signs off on this

File: src/types/utils.cpp [hunk 1]
```diff
@@ -318,6 +318,84 @@ catch (...)
...
+// - g - The green component of the color (0-100%).
+// - b - The blue component of the color (0-100%).
+// Return Value:
+// - The color defined by the given components.
+til::color Utils::ColorFromRGB100(const int r, const int g, const int b) noexcept
+{
+    // The color class is expecting components in the range 0 to 255,
+    // so we need to scale our percentage values by 255/100. We can
+    // optimise this conversion with a pre-created lookup table.
+    static constexpr auto scale100To255 = [] {
```
Comment (DHowett) [inline src/types/utils.cpp:334]: I love that this is compiled-in :D

File: src/types/utils.cpp [hunk 2]
```diff
@@ -318,6 +318,84 @@ catch (...)
...
+
+// Routine Description:
+// - Constructs a til::color value from HLS components.
+// Arguments:
+// - h - The hue component of the color (0-360°).
+// - l - The luminosity component of the color (0-100%).
+// - s - The saturation component of the color (0-100%).
+// Return Value:
+// - The color defined by the given components.
+til::color Utils::ColorFromHLS(const int h, const int l, const int s) noexcept
```
Comment (DHowett) [inline src/types/utils.cpp:357]: Eventually, I will file a workitem for moving this stuff into til/color.h itself (not necessarily into the `color` class, mind). We have about three copies of the HSL/HLS code, and perhaps we can get it down to 1. :smile:

**Review by DHowett — APPROVED** [review]
Lovely, as always. Thank you!

File: src/types/utils.cpp (hunk 2 above)
Comment (j4james) [reply src/types/utils.cpp:357]: Yeah, I'm totally on board with this. I was initially expecting to put it somewhere related to the `til::color` class - I only went with `Utils` because that's what we were doing for the XTerm color routines. But as you noted below, these conversions are somewhat unique, and would probably need to be named something like `ColorFromDecRGB` and `ColorFromDecHLS`.
